chips/cc26x2/src/uart.rs

```diff
@@ -160,39 +185,83 @@ impl UART {
     }
 
     fn enable_interrupts(&self) {
-        // Disable all UART interrupts
-        self.registers.imsc.modify(Interrupts::ALL_INTERRUPTS::SET);
+        // clear all
+        self.registers.icr.write(Interrupts::ALL_INTERRUPTS::Clear);
+        // set only interrupts used
+        self.registers.imsc.modify(
+            Interrupts::RX::SET
+                + Interrupts::RX_TIMEOUT::SET
+                + Interrupts::END_OF_TRANSMISSION::SET,
+        );
     }
 
     /// Clears all interrupts related to UART.
     pub fn handle_interrupt(&self) {
         // Clear interrupts
         self.registers.icr.write(Interrupts::ALL_INTERRUPTS::SET);
 
-        self.transaction.take().map(|mut transaction| {
-            transaction.index += 1;
-            if transaction.index < transaction.length {
-                self.send_byte(transaction.buffer[transaction.index]);
-                self.transaction.put(transaction);
-            } else {
+        self.rx.take().map(|mut rx| {
+            while self.rx_fifo_not_empty() && rx.index < rx.length {
+                let byte = self.read_byte();
+                rx.buffer[rx.index] = byte;
+                rx.index += 1;
+            }
+
+            if rx.index == rx.length {
                 self.client.map(move |client| {
-                    client.transmit_complete(
-                        transaction.buffer,
+                    client.receive_complete(
+                        rx.buffer,
+                        rx.index,
                         kernel::hil::uart::Error::CommandComplete,
                     );
                 });
+            } else {
+                self.rx.put(rx);
+            }
+        });
+        // if there is no client, empty the buffer into the void
+        if self.rx_fifo_not_empty() {
+            self.read_byte();
+        }
+
+        self.tx.take().map(|mut tx| {
+            // if a big buffer was given, this could be a very long call
+            if self.tx_fifo_not_full() && tx.index < tx.length {
+                self.send_byte(tx.buffer[tx.index]);
+                tx.index += 1;
+            }
+            if tx.index == tx.length {
+                self.client.map(move |client| {
+                    client.transmit_complete(tx.buffer, kernel::hil::uart::Error::CommandComplete);
+                });
+            } else {
+                self.tx.put(tx);
             }
         });
     }
 
-    /// Transmits a single byte if the hardware is ready.
+    // Pushes a byte into the TX FIFO.
+    #[inline]
```

Comment:
It doesn't hurt to have this, but I'm very surprised if this doesn't happen anyway.

Reply:
Agreed. C++ habit